src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUMatrixMemoryManager.java

```diff
@@ -94,10 +94,12 @@ else if(sparsePtr.val != null)
 	 * Get pointers from the first memory sections "Matrix Memory"
 	 * @param locked return locked pointers if true
 	 * @param dirty return dirty pointers if true
+	 * @param isCleanupEnabled return pointers marked for cleanup if true
 	 * @return set of pointers
 	 */
-	Set<Pointer> getPointers(boolean locked, boolean dirty) {
-		return gpuObjects.stream().filter(gObj -> gObj.isLocked() == locked && gObj.isDirty() == dirty).flatMap(gObj -> getPointers(gObj).stream()).collect(Collectors.toSet());
+	Set<Pointer> getPointers(boolean locked, boolean dirty, boolean isCleanupEnabled) {
+		return gpuObjects.stream().filter(gObj -> gObj.isLocked() == locked && gObj.isDirty() == dirty
+				&& gObj.mat.isCleanupEnabled() == isCleanupEnabled).flatMap(gObj -> getPointers(gObj).stream()).collect(Collectors.toSet());
```

Comment:
This will end up clearing pointers for pinned weight matrices as they are unlocked and not-dirty (since they're read only) at the end of a script invocation. This ends up causing a `CUDNN_STATUS_MAPPING_ERROR`. I think we should just never clean up pointers for pinned data (i.e. when cleanup is not enabled). We can fix this by either changing the above to:

```
gpuObjects.stream().filter(gObj -> gObj.isLocked() == locked && gObj.isDirty() == dirty || gObj.mat.isCleanupEnabled() == isCleanupEnabled)
```

Or by adding a separate check to see if a pointer being cleared is associated with a GPU object which is marked as pinned. My preference is for the latter as it seems easier to understand.